ocrd/ocrd/mets_server.py

```diff
@@ -0,0 +1,253 @@
+        self.log = getLogger('ocrd.mets_client.%s' % ('uds' if socket else 'tcp'))
+        if socket:
+            self.url = f'http+unix://{socket.replace("/", "%2F")}'
+            self.session = requests_unixsocket_session()
+        else:
+            self.url = f'http://{host}:{port}'
+            self.session = requests_session()
+
+    def find_files(self, **kwargs):
+        r = self.session.request('GET', self.url, params={**kwargs})
+        for f in r.json()['files']:
+            yield ClientSideOcrdFile(None, ID=f['file_id'], pageId=f['page_id'], fileGrp=f['file_grp'], url=f['url'], mimetype=f['mimetype'])
+
+    def find_all_files(self, *args, **kwargs):
+        return list(self.find_files(*args, **kwargs))
+
+    def add_agent(self, *args, **kwargs):
+        return self.session.request('POST', f'{self.url}/agent', data=OcrdAgentModel.create(**kwargs))
+
+    @property
+    def file_groups(self):
+        return self.session.request('GET', f'{self.url}/file_groups').json()['file_groups']
+
+    @deprecated_alias(pageId="page_id")
+    @deprecated_alias(ID="file_id")
+    def add_file(self, file_grp, content=None, file_id=None, url=None, mimetype=None, page_id=None, **kwargs):
+        r = self.session.request(
+            'POST',
+            self.url,
+            data=OcrdFileModel.create(
+                file_id=file_id,
+                file_grp=file_grp,
+                page_id=page_id,
+                mimetype=mimetype,
+                url=url).dict(),
+            files={'data': content}
+        )
+
+    def save(self):
+        self.session.request('PUT', self.url)
+
+
+#
+# Server
+#
+
+class OcrdMetsServer():
+
+    def __init__(self, workspace, host, port, socket):
+        self.workspace = workspace
+        self.host = host
+        self.port = port
+        self.socket = socket
+        self.log = getLogger('ocrd.workspace_client')
+
+
+    def startup(self):
+
+        # XXX HACK 
+        # circumventing dependency injection like this is bad and
+        # needs to be refactored once it's all runnign
+        workspace = self.workspace
+
+        app = FastAPI(
+            title="OCR-D METS Server",
+            description="Providing simultaneous write-access to mets.xml for OCR-D",
+        )
+
+        @app.exception_handler(ValidationError)
+        async def exception_handler_invalid400(request: Request, exc: ValidationError):
+            return JSONResponse(status_code=400, content=exc.errors())
+
+        @app.exception_handler(FileExistsError)
+        async def exception_handler_invalid400(request: Request, exc: FileExistsError):
+            return JSONResponse(status_code=400, content=str(exc))
+
+        @app.exception_handler(re.error)
+        async def exception_handler_invalid400(request: Request, exc: re.error):
+            return JSONResponse(status_code=400, content=f'invalid regex: {exc}')
+
+        @app.get("/", response_model=OcrdFileListModel)
```

Comment:
I personally prefer to get something like a "is running" response when executing a get to the server root (like if you just execute `ocrd workspace` you get usage information for example). I am unsure but maybe it is a good idea to move the find_files endpoint to `/files` (and add_file and related accordingly).

Reply:
Yeah, that makes sense. How about a common endpoint `/file`, which does search (`GET`) and file adding (`POST`)?

Reply:
9f27a98b0

Reply:
yes, `file` is better than `files`.